File: app-wrapper/scripts/build-backend.js
#!/usr/bin/env node
/**
 * PocketPaw Desktop — Build Script
 *
 * Prepares the backend bundle for packaging with electron-builder:
 *   1. Builds a Python wheel from the pocketpaw source (../pyproject.toml)
 *   2. Downloads the uv binary for the current platform
 *   3. Copies both into resources/backend/ for bundling
 *
 * Run: node scripts/build-backend.js
 */

import { execSync } from 'child_process'
import { existsSync, mkdirSync, copyFileSync, readdirSync, writeFileSync } from 'fs'
import { join, resolve } from 'path'
import { platform, arch } from 'os'

const ROOT = resolve(join(import.meta.dirname, '..'))
const PROJECT_ROOT = resolve(join(ROOT, '..'))
const RESOURCES_DIR = join(ROOT, 'resources')
const BACKEND_DIR = join(RESOURCES_DIR, 'backend')

function log(msg) {
  console.log(`  🔧 ${msg}`)
}

function ensureDir(dir) {
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true })
  }
}

// ─── Step 1: Build the pocketpaw wheel ────────────────────────────

function buildWheel() {
  log('Building pocketpaw wheel...')
  const distDir = join(PROJECT_ROOT, 'dist')

  try {
    // Use uv build to create the wheel (fastest)
    execSync('uv build --wheel', {
      cwd: PROJECT_ROOT,
      stdio: 'pipe',
      timeout: 120000
    })
  } catch {
    // Fallback to pip build
    log('uv build failed, trying pip...')
    execSync('python -m build --wheel', {
      cwd: PROJECT_ROOT,
      stdio: 'pipe',
      timeout: 120000
    })
  }

  // Find the built wheel
  const wheels = readdirSync(distDir).filter(f => f.endsWith('.whl'))
  if (wheels.length === 0) {
    throw new Error('No wheel found in dist/ after build')
  }

  // Copy the latest wheel
  const wheel = wheels.sort().pop()
  const src = join(distDir, wheel)
  const dst = join(BACKEND_DIR, wheel)
  ensureDir(BACKEND_DIR)
  copyFileSync(src, dst)
  log(`Wheel: ${wheel}`)

  return wheel
}

// ─── Step 2: Download uv binary ───────────────────────────────────

function downloadUv() {
  const os = platform()
  const cpu = arch()

  // Map to uv target names
  const targetMap = {
    'win32-x64': 'x86_64-pc-windows-msvc.zip',
    'win32-arm64': 'aarch64-pc-windows-msvc.zip',
    'darwin-x64': 'x86_64-apple-darwin.tar.gz',
    'darwin-arm64': 'aarch64-apple-darwin.tar.gz',
    'linux-x64': 'x86_64-unknown-linux-musl.tar.gz',
    'linux-arm64': 'aarch64-unknown-linux-musl.tar.gz'
  }

  const target = targetMap[`${os}-${cpu}`]
  if (!target) {
    log(`⚠️  No uv binary for ${os}-${cpu}, skipping (will use system uv)`)
    return
  }

  const uvDir = join(BACKEND_DIR, 'uv')
  ensureDir(uvDir)

  const isWin = os === 'win32'
  const uvBin = join(uvDir, isWin ? 'uv.exe' : 'uv')

  if (existsSync(uvBin)) {
    log('uv binary already exists, skipping download')
    return
  }

  // Use the pinned version from bootstrap.py
  const version = '0.6.6'
  const url = `https://github.com/astral-sh/uv/releases/download/${version}/uv-${target}`

  log(`Downloading uv ${version} for ${os}-${cpu}...`)

  try {
    if (target.endsWith('.zip')) {
      // Download and extract zip (Windows)
      const tmpZip = join(BACKEND_DIR, 'uv-tmp.zip')
      execSync(`curl -fsSL "${url}" -o "${tmpZip}"`, { timeout: 60000, stdio: 'pipe' })

      // Extract uv.exe from the zip
      execSync(`tar -xf "${tmpZip}" -C "${uvDir}" --strip-components=1`, {
        timeout: 30000, stdio: 'pipe'
      })

      // Clean up
      try { require('fs').unlinkSync(tmpZip) } catch {}
    } else {
      // tar.gz (macOS/Linux)
      execSync(
        `curl -fsSL "${url}" | tar xz -C "${uvDir}" --strip-components=1`,
        { timeout: 60000, stdio: 'pipe', shell: true }
      )
    }

    if (existsSync(uvBin)) {
      log(`uv binary: ${uvBin}`)
    } else {
      log('⚠️  uv download completed but binary not found')
    }
  } catch (err) {
    log(`⚠️  Failed to download uv: ${err.message}`)
  }
}

// ─── Step 3: Write bootstrap metadata ─────────────────────────────

function writeMetadata(wheelName) {
  const meta = {
    version: JSON.parse(
      require('fs').readFileSync(join(ROOT, 'package.json'), 'utf-8')
    ).version,
    wheel: wheelName,
    builtAt: new Date().toISOString(),
    platform: `${platform()}-${arch()}`
  }

  writeFileSync(
    join(BACKEND_DIR, 'bundle.json'),
    JSON.stringify(meta, null, 2)
  )
  log(`Metadata written to bundle.json`)
}

// ─── Main ─────────────────────────────────────────────────────────

try {
  log('Preparing backend bundle...')
  ensureDir(BACKEND_DIR)

  const wheel = buildWheel()
  downloadUv()
  writeMetadata(wheel)

  log('✅ Backend bundle ready in resources/backend/')
} catch (err) {
  console.error(`\n  ❌ Build failed: ${err.message}\n`)
  process.exit(1)
}

File: app-wrapper/scripts/build-backend.py
#!/usr/bin/env python3
"""
PocketPaw Desktop — Nuitka Build Script

Compiles the PocketPaw Python backend into a standalone binary using Nuitka.
The resulting binary is placed in app-wrapper/resources/backend/ for bundling
with Electron via electron-builder.

Usage:
    python scripts/build-backend.py           # Build for current platform
    python scripts/build-backend.py --onefile # Single-file binary

Prerequisites:
    pip install nuitka
    # Windows: Visual Studio Build Tools
    # macOS:   xcode-select --install
    # Linux:   sudo apt install gcc g++
"""

import platform
import subprocess
import sys
from pathlib import Path

APP_NAME = "pocketpaw-server"
ENTRY = "src/pocketpaw/__main__.py"

# Resolve paths relative to the project root
PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent  # pocketpaw-/
APP_WRAPPER = PROJECT_ROOT / "app-wrapper"
OUTPUT_DIR = APP_WRAPPER / "resources" / "backend"

# PocketPaw version from pyproject.toml (fallback)
try:
    import tomllib
    with open(PROJECT_ROOT / "pyproject.toml", "rb") as f:
        pyproject = tomllib.load(f)
    VERSION = pyproject.get("project", {}).get("version", "0.4.9")
except Exception:
    VERSION = "0.4.9"


def build():
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    cmd = [
        sys.executable, "-m", "nuitka",
        "--standalone",
        "--enable-plugin=anti-bloat",
        "--assume-yes-for-downloads",
        f"--output-filename={APP_NAME}",
        f"--product-version={VERSION}",
        "--company-name=PocketPaw",
        f"--product-name={APP_NAME}",
        "--file-description=PocketPaw AI Agent Server",
        f"--copyright=Copyright 2026 PocketPaw",
        f"--output-dir={OUTPUT_DIR}",
    ]

    # ── Include PocketPaw packages ────────────────────────────────
    # Core packages that Nuitka needs to find
    packages = [
        "pocketpaw",
        "uvicorn",
        "fastapi",
        "starlette",
        "pydantic",
        "anyio",
        "httpx",
        "jinja2",
        "aiofiles",
        "rich",
        "toml",
        "yaml",
        "dotenv",
        "multipart",
        "websockets",
    ]
    for pkg in packages:
        cmd.append(f"--include-package={pkg}")

    # ── Include data directories (templates, static, css, js) ────
    frontend_dir = PROJECT_ROOT / "src" / "pocketpaw" / "frontend"
    if frontend_dir.exists():
        for subdir in ["templates", "css", "js"]:
            src = frontend_dir / subdir
            if src.exists():
                cmd.append(f"--include-data-dir={src}=pocketpaw/frontend/{subdir}")

    # ── Exclude test/dev modules to reduce binary size ───────────
    cmd.extend([
        "--nofollow-import-to=pytest,unittest,test,setuptools,pip,tkinter",
        "--nofollow-import-to=sphinx,docutils,pygments",
    ])

    # ── Platform-specific flags ──────────────────────────────────
    if platform.system() == "Windows":
        cmd.extend([
            "--windows-disable-console",  # No console window
        ])
        icon = APP_WRAPPER / "resources" / "icon.ico"
        if icon.exists():
            cmd.append(f"--windows-icon-from-ico={icon}")

    elif platform.system() == "Darwin":
        icon = APP_WRAPPER / "resources" / "icon.icns"
        if icon.exists():
            cmd.append(f"--macos-app-icon={icon}")

    # ── Onefile option ───────────────────────────────────────────
    if "--onefile" in sys.argv:
        cmd.append("--onefile")

    # ── Speed up rebuilds ────────────────────────────────────────
    cmd.append("--ccache")

    # Entry point
    cmd.append(str(PROJECT_ROOT / ENTRY))

    print(f"\n  🔧 Building {APP_NAME} v{VERSION}...")
    print(f"  📁 Output: {OUTPUT_DIR}")
    print(f"  🖥️  Platform: {platform.system()} {platform.machine()}")
    print(f"\n  Command: {' '.join(str(c) for c in cmd)}\n")

    result = subprocess.run(cmd)
    if result.returncode == 0:
        # Find the output binary
        if platform.system() == "Windows":
            binary = OUTPUT_DIR / f"{APP_NAME}.exe"
        else:
            binary = OUTPUT_DIR / APP_NAME

        if binary.exists():
            size_mb = binary.stat().st_size / (1024 * 1024)
            print(f"\n  ✅ Build complete: {binary} ({size_mb:.1f} MB)\n")
        else:
            # Standalone mode creates a .dist directory
            dist_dir = OUTPUT_DIR / f"{APP_NAME}.dist"
            if dist_dir.exists():
                print(f"\n  ✅ Build complete: {dist_dir}\n")
            else:
                print(f"\n  ⚠️  Build completed but binary not found at expected path\n")
    else:
        print(f"\n  ❌ Build failed with exit code {result.returncode}\n")
        sys.exit(1)


if __name__ == "__main__":
    build()
